File: grpc-trading-platform/user-service/src/main/java/org/hle/user/service/UserService.java
package org.hle.user.service;

import org.hle.user.*;
import org.hle.user.service.handler.StockTradeRequestHandler;
import org.hle.user.service.handler.UserInformationRequestHandler;
import io.grpc.stub.StreamObserver;
import net.devh.boot.grpc.server.service.GrpcService;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

@GrpcService
public class UserService extends UserServiceGrpc.UserServiceImplBase {

    private static final Logger log = LoggerFactory.getLogger(UserService.class);

    private final UserInformationRequestHandler userRequestHandler;
    private final StockTradeRequestHandler tradeRequestHandler;

    public UserService(UserInformationRequestHandler userRequestHandler, StockTradeRequestHandler tradeRequestHandler) {
        this.userRequestHandler = userRequestHandler;
        this.tradeRequestHandler = tradeRequestHandler;
    }

    @Override
    public void getUserInformation(UserInformationRequest request, StreamObserver<UserInformation> responseObserver) {
        log.info("user information for id {}", request.getUserId());
        var userInformation = this.userRequestHandler.getUserInformation(request);
        responseObserver.onNext(userInformation);
        responseObserver.onCompleted();
    }

    @Override
    public void tradeStock(StockTradeRequest request, StreamObserver<StockTradeResponse> responseObserver) {
        log.info("{}", request);
        var response = TradeAction.SELL.equals(request.getAction())
                ? tradeRequestHandler.sellStock(request)
                : tradeRequestHandler.buyStock(request);

        responseObserver.onNext(response);
        responseObserver.onCompleted();
    }
}

File: grpc-trading-official/user-service/src/main/kotlin/org/hle/userservice/service/UserService.kt
package org.hle.userservice.service

import org.hle.user.UserServiceGrpc.UserServiceImplBase
import org.hle.user.UserInformationRequest
import org.hle.user.UserInformation
import org.hle.user.StockTradeRequest
import org.hle.user.StockTradeResponse
import org.hle.user.TradeAction
import org.hle.userservice.handler.UserInformationRequestHandler
import org.hle.userservice.handler.StockTradeRequestHandler
import org.springframework.stereotype.Service

@Service
class UserService(private val userRequestHandler: UserInformationRequestHandler,
                  private val tradeRequestHandler: StockTradeRequestHandler
) :
    UserServiceImplBase() {

    override fun getUserInformation(
        request: UserInformationRequest,
        responseObserver: io.grpc.stub.StreamObserver<UserInformation?>
    ) {
        log.info("user information for id {}", request.getUserId())
        val userInformation: UserInformation =
            this.userRequestHandler.getUserInformation(request)
        responseObserver.onNext(userInformation)
        responseObserver.onCompleted()
    }

    override fun tradeStock(
        request: StockTradeRequest,
        responseObserver: io.grpc.stub.StreamObserver<StockTradeResponse?>
    ) {
        log.info("{}", request)
        val response: StockTradeResponse =
            if (TradeAction.SELL == request.getAction())
                tradeRequestHandler.sellStock(request)
            else
                tradeRequestHandler.buyStock(request)

        responseObserver.onNext(response)
        responseObserver.onCompleted()
    }

    companion object {
        private val log: org.slf4j.Logger = org.slf4j.LoggerFactory.getLogger(UserService::class.java)
    }
}
